lib/cartopy/mpl/patch.py

```diff
@@ -170,12 +175,26 @@ def path_to_geos(path, force_ccw=False):
         # lake within a land mass).  Sometimes there is a further geom within
         # this interior (e.g. an island in a lake, or some instances of
         # contours).  This needs to be a new external geom in the collection.
-        if geom.is_empty:
-            pass
-        elif (len(collection) > 0 and
+        try:
+            is_inside = (len(collection) > 0 and
+                isinstance(collection[-1][0], sgeom.Polygon) and
+                isinstance(geom, sgeom.Polygon) and
+                collection[-1][0].contains(geom.exterior))
+        except GEOSException:
+            # If the GEOSException is raised, it is likely that the polygon
+            # is invalid.  In this case, we can't use the contains method.
+            # Therefore, we need to perform a repair on the Polygon object
+            # and then attempt the contains method again. Related Issue: #2370
+            polygon = collection[-1][0].buffer(0)
+            collection[-1] = (polygon, collection[-1][1])
+            is_inside = (len(collection) > 0 and
                 isinstance(collection[-1][0], sgeom.Polygon) and
                 isinstance(geom, sgeom.Polygon) and
-                collection[-1][0].contains(geom.exterior)):
+                collection[-1][0].contains(geom.exterior))
+
+        if geom.is_empty:
+            pass
```

Comment:
Since this check is cheap compared with the contains method, I think it would be good to keep it first and only define `is_inside` if `geom` is not empty.